### PR TITLE
[hold_worker] add a worker manager

### PR DESCRIPTION
**Pull request by rajdeepv**
- Add option to keep some workers alive while others are busy. These are for backup in case busy workers malfunction
- Use a worker manager to assign tasks to worker

### CODE REVIEW

File: lib/parallel_cucumber/worker_manager.rb [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+    end
+
+    def create_workers(number_of_workers)
+      number_of_workers.times do |index|
+        @workers["W#{index}"] =
+          ParallelCucumber::Worker.new(options: @options, index: index, stdout_logger: @logger, manager: self)
+      end
+    end
+
+    def start_managing
```
Comment (fr0l) [inline lib/parallel_cucumber/worker_manager.rb]: private

File: lib/parallel_cucumber/worker_manager.rb [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+      @healthy_workers = Queue.new
+    end
+
+    def start(number_of_workers)
+      create_workers(number_of_workers)
+      start_managing
+      start_workers
+    end
+
+    def create_workers(number_of_workers)
```
Comment (fr0l) [inline lib/parallel_cucumber/worker_manager.rb]: private

File: lib/parallel_cucumber/worker_manager.rb [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+          else
+            kill_all_workers
+            break
+          end
+        end
+        @finished = true
+      end
+    end
+
+    def start_workers
```
Comment (fr0l) [inline lib/parallel_cucumber/worker_manager.rb]: private

File: lib/parallel_cucumber/worker_manager.rb [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+      @unchecked_workers.enq(worker)
+    end
+
+    private
+
+    def kill_all_workers
+      @workers.values.each { |w| w.assign_job(Job.new(Job::DIE)) }
+    end
+
+    def kill_needless_workers
```
Comment (fr0l) [inline lib/parallel_cucumber/worker_manager.rb]: kill_surplus_workers

File: lib/parallel_cucumber/cli.rb [hunk 5]
```diff
@@ -19,7 +19,8 @@ class Cli
       n: 0, # Default: computed from longest list in json parameters, minimum 1.
       queue_connection_params: ['redis://127.0.0.1:6379', DateTime.now.strftime('queue-%Y%m%d%H%M%S')],
       worker_delay: 0,
-      test_command: 'cucumber'
+      test_command: 'cucumber',
+      backup_worker_count: 1
     }.freeze
```
Comment (bayandin) [inline lib/parallel_cucumber/cli.rb]: Make it 0 to keep default behaviour the same as before

File: lib/parallel_cucumber/job.rb [hunk 6]
```diff
@@ -0,0 +1,13 @@
+module ParallelCucumber
+  class Job
+    RUN_TESTS = 'run_tests'.freeze
+    PRECHECK = 'precheck'.freeze
+    DIE = 'die'.freeze
```
Comment (bayandin) [inline lib/parallel_cucumber/job.rb]: Maybe use symbols? I don't mind to have strings anyway..

File: lib/parallel_cucumber/worker_manager.rb (hunk 1 above)
Comment (rajdeepv) [reply lib/parallel_cucumber/worker_manager.rb]: done

File: lib/parallel_cucumber/worker_manager.rb (hunk 2 above)
Comment (rajdeepv) [reply lib/parallel_cucumber/worker_manager.rb]: done

File: lib/parallel_cucumber/worker_manager.rb (hunk 4 above)
Comment (rajdeepv) [reply lib/parallel_cucumber/worker_manager.rb]: done

File: Gemfile.lock [hunk 7]
```diff
@@ -1,7 +1,7 @@
 PATH
   remote: .
   specs:
-    parallel_cucumber (0.2.16)
+    parallel_cucumber (0.2.17pre3)
```
[the remark this answers is not in the document]
Comment (rajdeepv) [reply Gemfile.lock]: done

File: lib/parallel_cucumber.rb [hunk 8]
```diff
@@ -11,4 +11,6 @@
 require 'parallel_cucumber/main'
 require 'parallel_cucumber/status'
 require 'parallel_cucumber/version'
+require 'parallel_cucumber/worker_manager'
 require 'parallel_cucumber/worker'
+require 'parallel_cucumber/job'
```
[the remark this answers is not in the document]
Comment (rajdeepv) [reply lib/parallel_cucumber.rb]: done

File: lib/parallel_cucumber/job.rb (hunk 6 above)
Comment (rajdeepv) [reply lib/parallel_cucumber/job.rb]: changed to symbols

File: lib/parallel_cucumber/cli.rb (hunk 5 above)
Comment (rajdeepv) [reply lib/parallel_cucumber/cli.rb]: changed

File: lib/parallel_cucumber/worker_manager.rb (hunk 3 above)
Comment (rajdeepv) [reply lib/parallel_cucumber/worker_manager.rb]: done

**Pull request merged**